

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API examples &mdash; aide 1.1.01 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=e50f1ef7"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Data Structures" href="data_structures.html" />
    <link rel="prev" title="Installing AIDE" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            aide
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installing AIDE</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#checking-which-protein-models-are-available-given-the-data-you-have">1. Checking which protein models are available given the data you have</a></li>
<li class="toctree-l2"><a class="reference internal" href="#in-silico-mutagenesis-using-msatransformer">2. In silico mutagenesis using MSATransformer</a></li>
<li class="toctree-l2"><a class="reference internal" href="#compare-a-couple-of-zero-shot-predictors-against-experimental-data">3. Compare a couple of zero shot predictors against experimental data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#train-a-supervised-model-to-predict-activity-on-an-experimental-combinatorial-library-test-on-sequences-with-greater-mutational-depth-than-training">4. Train a supervised model to predict activity on an experimental combinatorial library, test on sequences with greater mutational depth than training</a></li>
<li class="toctree-l2"><a class="reference internal" href="#train-a-supervised-predictor-on-a-set-of-homologs-focusing-only-on-positions-of-known-importance-wrap-the-entire-process-into-an-sklearn-pipeline-including-some-standard-sklearn-transormers-and-make-predictions-for-a-new-set-of-homologs">5. Train a supervised predictor on a set of homologs, focusing only on positions of known importance, wrap the entire process into an sklearn pipeline including some standard sklearn transormers, and make predictions for a new set of homologs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#create-new-embedder-or-predictor-within-the-aide-framework">6. Create new embedder or predictor within the aide framework</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="data_structures.html">Data Structures</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_compatibility.html">Model Compatibility</a></li>
<li class="toctree-l1"><a class="reference internal" href="protein_model.html">ProteinModelWrapper</a></li>
<li class="toctree-l1"><a class="reference internal" href="zero_shot.html">Zero-Shot Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="supervised.html">Supervised Learning</a></li>
<li class="toctree-l1"><a class="reference internal" href="saturation_mutagenesis.html">Saturation Mutagenesis</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipelines.html">Building ML Pipelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="caching.html">Caching Model Outputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="position_specific.html">Position-Specific Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing_models.html">Contributing Models to AIDE</a></li>
<li class="toctree-l1"><a class="reference internal" href="structure_pred.html">Structure Prediction with SoloSeq</a></li>
<li class="toctree-l1"><a class="reference internal" href="msa_search.html">Generating MSAs with MMseqs2</a></li>
<li class="toctree-l1"><a class="reference internal" href="badass.html">Protein Optimization with BADASS</a></li>
<li class="toctree-l1"><a class="reference internal" href="roadmap.html">Roadmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="resource_test.html">Resource testing</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules.html">aide_predict</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">aide</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">API examples</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/user_guide/api_examples.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section class="tex2jax_ignore mathjax_ignore" id="api-examples">
<h1>API examples<a class="headerlink" href="#api-examples" title="Link to this heading"></a></h1>
<p>The following should look and feel like canonical sklearn tasks/code. See the <code class="docutils literal notranslate"><span class="pre">demo</span></code> folder for more details and executable examples. Also see the <a class="reference external" href="https://colab.research.google.com/drive/1baz4DdYkxaw6pPRTDscwh2o-Xqum5Krp#scrollTo=AV9VXhM6ebgI">colab notebook</a> to play with some if its capabilities in the cloud. Finally, checkout the notebooks in <code class="docutils literal notranslate"><span class="pre">showcase</span></code> where we conduct two full protein predictions optimization and scoring tasks on real data that are greater than small example sets.</p>
<section id="checking-which-protein-models-are-available-given-the-data-you-have">
<h2>1. Checking which protein models are available given the data you have<a class="headerlink" href="#checking-which-protein-models-are-available-given-the-data-you-have" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">aide_predict.utils.checks</span> <span class="kn">import</span> <span class="n">check_model_compatability</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span><span class="s1">&#39;csv_file.csv&#39;</span><span class="p">,</span> <span class="n">seq_col</span><span class="o">=</span><span class="s1">&#39;sequence&#39;</span><span class="p">)</span>
<span class="n">wt</span> <span class="o">=</span> <span class="n">seqs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="n">check_model_compatibility</span><span class="p">(</span>
    <span class="n">training_msa</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">training_sequences</span><span class="o">=</span><span class="n">seqs</span><span class="p">,</span>
    <span class="n">wt</span><span class="o">=</span><span class="n">wt</span><span class="p">,</span>
<span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span><span class="p">{</span><span class="s1">&#39;compatible&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;ESM2Embedding&#39;</span><span class="p">,</span>
  <span class="s1">&#39;ESM2LikelihoodWrapper&#39;</span><span class="p">,</span>
  <span class="s1">&#39;KmerEmbedding&#39;</span><span class="p">,</span>
  <span class="s1">&#39;OneHotProteinEmbedding&#39;</span><span class="p">],</span>
 <span class="s1">&#39;incompatible&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;EVMutationWrapper&#39;</span><span class="p">,</span>
  <span class="s1">&#39;HMMWrapper&#39;</span><span class="p">,</span>
  <span class="s1">&#39;MSATransformerEmbedding&#39;</span><span class="p">,</span>
  <span class="s1">&#39;MSATransformerLikelihoodWrapper&#39;</span><span class="p">,</span>
  <span class="s1">&#39;OneHotAlignedEmbedding&#39;</span><span class="p">,</span>
  <span class="s1">&#39;SaProtEmbedding&#39;</span><span class="p">,</span>
  <span class="s1">&#39;SaProtLikelihoodWrapper&#39;</span><span class="p">,</span>
  <span class="s1">&#39;VESPAWrapper&#39;</span><span class="p">]}</span>
</pre></div>
</div>
</section>
<section id="in-silico-mutagenesis-using-msatransformer">
<h2>2. In silico mutagenesis using MSATransformer<a class="headerlink" href="#in-silico-mutagenesis-using-msatransformer" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># data preparation</span>
<span class="n">wt</span> <span class="o">=</span> <span class="n">ProteinSequence</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="s2">&quot;data/msa.fasta&quot;</span><span class="p">)</span> <span class="c1"># assigns the msa attribute of the sequence</span>
<span class="n">wt</span><span class="o">.</span><span class="n">has_msa</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kc">True</span>
<span class="n">wt</span><span class="o">.</span><span class="n">msa_same_width</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kc">True</span>
<span class="n">library</span> <span class="o">=</span> <span class="n">wt</span><span class="o">.</span><span class="n">saturation_mutagenesis</span><span class="p">()</span>
<span class="n">mutations</span> <span class="o">=</span> <span class="n">library</span><span class="o">.</span><span class="n">ids</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mutations</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="o">&gt;&gt;&gt;</span> <span class="s1">&#39;L1A&#39;</span>

<span class="c1"># model fitting</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">MSATransformerLikelihoodWrapper</span><span class="p">(</span>
   <span class="n">wt</span><span class="o">=</span><span class="n">wt</span><span class="p">,</span>
   <span class="n">marginal_method</span><span class="o">=</span><span class="s2">&quot;masked_marginal&quot;</span>
<span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>

<span class="c1"># make predictions for each mutated sequence</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">library</span><span class="p">)</span>

<span class="n">results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;mutation&#39;</span><span class="p">:</span> <span class="n">mutations</span><span class="p">,</span> <span class="s1">&#39;seqeunce&#39;</span><span class="p">:</span> <span class="n">library</span><span class="p">,</span><span class="s1">&#39;prediction&#39;</span><span class="p">:</span> <span class="n">predictions</span><span class="p">})</span>
</pre></div>
</div>
</section>
<section id="compare-a-couple-of-zero-shot-predictors-against-experimental-data">
<h2>3. Compare a couple of zero shot predictors against experimental data<a class="headerlink" href="#compare-a-couple-of-zero-shot-predictors-against-experimental-data" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># data preparation</span>
<span class="n">X</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span><span class="s2">&quot;data/experimental_data.csv&quot;</span><span class="p">,</span> <span class="n">seq_col</span><span class="o">=</span><span class="s1">&#39;sequence&#39;</span><span class="p">,</span> <span class="n">id_col</span><span class="o">=</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="n">label_cols</span><span class="o">=</span><span class="s1">&#39;experimental_value&#39;</span><span class="p">)</span>
<span class="n">wt</span> <span class="o">=</span> <span class="n">X</span><span class="p">[</span><span class="s1">&#39;my_id_for_WT&#39;</span><span class="p">]</span>
<span class="n">msa</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="s2">&quot;data/msa.fasta&quot;</span><span class="p">)</span>
<span class="n">wt</span><span class="o">.</span><span class="n">msa</span> <span class="o">=</span> <span class="n">msa</span>

<span class="c1"># model defenitions</span>
<span class="n">evmut</span> <span class="o">=</span> <span class="n">EVMutation</span><span class="p">(</span><span class="n">wt</span><span class="o">=</span><span class="n">wt</span><span class="p">,</span> <span class="n">metadata_folder</span><span class="o">=</span><span class="s1">&#39;./tmp/evm&#39;</span><span class="p">)</span>
<span class="n">evmut</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="n">esm2</span> <span class="o">=</span> <span class="n">ESM2LikelihoodWrapper</span><span class="p">(</span><span class="n">wt</span><span class="o">=</span><span class="n">wt</span><span class="p">,</span> <span class="n">model_checkpoint</span><span class="o">=</span><span class="s1">&#39;esm2_t33_650M_UR50S&#39;</span><span class="p">)</span>
<span class="n">esm2</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="n">models</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;evmut&#39;</span><span class="p">:</span> <span class="n">evmut</span><span class="p">,</span> <span class="s1">&#39;esm2&#39;</span><span class="p">:</span> <span class="n">esm2</span><span class="p">}</span>

<span class="c1"># model fitting and scoring</span>
<span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">models</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">score</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> score: </span><span class="si">{</span><span class="n">score</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="train-a-supervised-model-to-predict-activity-on-an-experimental-combinatorial-library-test-on-sequences-with-greater-mutational-depth-than-training">
<h2>4. Train a supervised model to predict activity on an experimental combinatorial library, test on sequences with greater mutational depth than training<a class="headerlink" href="#train-a-supervised-model-to-predict-activity-on-an-experimental-combinatorial-library-test-on-sequences-with-greater-mutational-depth-than-training" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># data preparation</span>
<span class="n">sequences</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span><span class="s2">&quot;data/experimental_data.csv&quot;</span><span class="p">,</span> <span class="n">seq_col</span><span class="o">=</span><span class="s1">&#39;sequence&#39;</span><span class="p">,</span> <span class="n">id_col</span><span class="o">=</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="n">label_cols</span><span class="o">=</span><span class="s1">&#39;experimental_value&#39;</span><span class="p">)</span>
<span class="n">sequences</span><span class="o">.</span><span class="n">aligned</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kc">True</span>
<span class="n">sequences</span><span class="o">.</span><span class="n">fixed_length</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kc">True</span>

<span class="n">wt</span> <span class="o">=</span> <span class="n">sequences</span><span class="p">[</span><span class="s1">&#39;my_id_for_WT&#39;</span><span class="p">]</span>
<span class="n">mutational_depth</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">mutated_positions</span><span class="p">(</span><span class="n">wt</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">])</span>
<span class="n">test_mask</span> <span class="o">=</span> <span class="n">mutational_depth</span> <span class="o">&gt;</span> <span class="mi">5</span>
<span class="n">train_X</span> <span class="o">=</span> <span class="n">sequences</span><span class="p">[</span><span class="o">~</span><span class="n">test_mask</span><span class="p">]</span>
<span class="n">train_y</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="o">~</span><span class="n">test_mask</span><span class="p">]</span>
<span class="n">test_X</span> <span class="o">=</span> <span class="n">sequences</span><span class="p">[</span><span class="n">test_mask</span><span class="p">]</span>
<span class="n">test_y</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="n">test_mask</span><span class="p">]</span>

<span class="c1"># embeddings protein sequences</span>
<span class="c1"># use mean pool embeddings of esm2</span>
<span class="n">embedder</span> <span class="o">=</span> <span class="n">ESM2Embedding</span><span class="p">(</span><span class="n">pool</span><span class="o">=</span><span class="s1">&#39;mean&#39;</span><span class="p">)</span>
<span class="n">train_X</span> <span class="o">=</span> <span class="n">embedder</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">train_X</span><span class="p">)</span>
<span class="n">test_X</span> <span class="o">=</span> <span class="n">embedder</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">test_X</span><span class="p">)</span>

<span class="c1"># model fitting</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">RandomForestRegressor</span><span class="p">()</span>
<span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">train_X</span><span class="p">,</span> <span class="n">train_y</span><span class="p">)</span>

<span class="c1"># model scoring</span>
<span class="n">train_score</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">train_X</span><span class="p">,</span> <span class="n">train_y</span><span class="p">)</span>
<span class="n">test_score</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">test_X</span><span class="p">,</span> <span class="n">test_y</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Train score: </span><span class="si">{</span><span class="n">train_score</span><span class="si">}</span><span class="s2">, Test score: </span><span class="si">{</span><span class="n">test_score</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="train-a-supervised-predictor-on-a-set-of-homologs-focusing-only-on-positions-of-known-importance-wrap-the-entire-process-into-an-sklearn-pipeline-including-some-standard-sklearn-transormers-and-make-predictions-for-a-new-set-of-homologs">
<h2>5. Train a supervised predictor on a set of homologs, focusing only on positions of known importance, wrap the entire process into an sklearn pipeline including some standard sklearn transormers, and make predictions for a new set of homologs<a class="headerlink" href="#train-a-supervised-predictor-on-a-set-of-homologs-focusing-only-on-positions-of-known-importance-wrap-the-entire-process-into-an-sklearn-pipeline-including-some-standard-sklearn-transormers-and-make-predictions-for-a-new-set-of-homologs" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># data preparation</span>
<span class="n">sequences</span><span class="p">,</span> <span class="n">y_train</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span><span class="s2">&quot;data/training_data.csv&quot;</span><span class="p">,</span> <span class="n">seq_col</span><span class="o">=</span><span class="s1">&#39;sequence&#39;</span><span class="p">,</span> <span class="n">id_col</span><span class="o">=</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="n">label_cols</span><span class="o">=</span><span class="s1">&#39;experimental_value&#39;</span><span class="p">)</span>

<span class="n">wt</span> <span class="o">=</span> <span class="n">sequences</span><span class="p">[</span><span class="s1">&#39;my_id_for_WT&#39;</span><span class="p">]</span>
<span class="n">wt_important_positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">20</span><span class="p">,</span> <span class="mi">21</span><span class="p">,</span> <span class="mi">22</span><span class="p">,</span> <span class="mi">33</span><span class="p">,</span> <span class="mi">45</span><span class="p">])</span> <span class="c1"># zero indexed, known from analysis elsewhere</span>
<span class="n">sequences</span><span class="o">.</span><span class="n">aligned</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kc">False</span>
<span class="n">sequences</span><span class="o">.</span><span class="n">fixed_length</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kc">False</span>

<span class="c1"># align the training sequences and get the important positions</span>
<span class="n">msa</span> <span class="o">=</span> <span class="n">sequences</span><span class="o">.</span><span class="n">align_all</span><span class="p">()</span>
<span class="n">msa</span><span class="o">.</span><span class="n">fixed_length</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kc">False</span>
<span class="n">msa</span><span class="o">.</span><span class="n">aligned</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kc">True</span>

<span class="n">wt_alignment_mapping</span> <span class="o">=</span> <span class="n">msa</span><span class="o">.</span><span class="n">get_alignment_mapping</span><span class="p">()[</span><span class="s1">&#39;my_id_for_WT&#39;</span><span class="p">]</span>
<span class="n">aligned_important_positions</span> <span class="o">=</span> <span class="n">wt_alignment_mapping</span><span class="p">[</span><span class="n">wt_important_positions</span><span class="p">]</span>

<span class="c1"># model defenitions</span>
<span class="n">embedder</span> <span class="o">=</span> <span class="n">OneHotAlignedEmbedding</span><span class="p">(</span><span class="n">important_positions</span><span class="o">=</span><span class="n">aligned_important_positions</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span>
<span class="n">scaler</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span>
<span class="n">feature_selector</span> <span class="o">=</span> <span class="n">VarianceThreshold</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
<span class="n">predictor</span> <span class="o">=</span> <span class="n">RandomForestRegressor</span><span class="p">()</span>
<span class="n">pipeline</span> <span class="o">=</span> <span class="n">Pipeline</span><span class="p">([</span>
    <span class="p">(</span><span class="s1">&#39;embedder&#39;</span><span class="p">,</span> <span class="n">embedder</span><span class="p">),</span>
    <span class="p">(</span><span class="s1">&#39;scaler&#39;</span><span class="p">,</span> <span class="n">scaler</span><span class="p">),</span>
    <span class="p">(</span><span class="s1">&#39;feature_selector&#39;</span><span class="p">,</span> <span class="n">feature_selector</span><span class="p">),</span>
    <span class="p">(</span><span class="s1">&#39;predictor&#39;</span><span class="p">,</span> <span class="n">predictor</span><span class="p">)</span>
<span class="p">])</span>

<span class="c1"># model fitting</span>
<span class="n">pipeline</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">y_train</span><span class="p">)</span>

<span class="c1"># score new unaligned homologs</span>
<span class="n">new_homologs</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="s2">&quot;data/new_homologs.fasta&quot;</span><span class="p">)</span>
<span class="n">y_pred</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">new_homologs</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="create-new-embedder-or-predictor-within-the-aide-framework">
<h2>6. Create new embedder or predictor within the aide framework<a class="headerlink" href="#create-new-embedder-or-predictor-within-the-aide-framework" title="Link to this heading"></a></h2>
<p>Here we create a K-mer counting embedding, except use Foldseek structure tokens instead of amino acids. We set the <code class="docutils literal notranslate"><span class="pre">_available</span></code> attribute to allow aide to dynamically check if the model is available to call.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Optional</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>

<span class="kn">from</span> <span class="nn">aide_predict.bespoke_models.base</span> <span class="kn">import</span> <span class="n">ProteinModelWrapper</span><span class="p">,</span> <span class="n">CanHandleAlignedSequencesMixin</span><span class="p">,</span> <span class="n">RequiresStructureMixin</span>
<span class="kn">from</span> <span class="nn">aide_predict.utils.data_structures</span> <span class="kn">import</span> <span class="n">ProteinSequences</span><span class="p">,</span> <span class="n">ProteinSequence</span>
<span class="kn">from</span> <span class="nn">aide_predict.utils.common</span> <span class="kn">import</span> <span class="n">MessageBool</span>
<span class="kn">from</span> <span class="nn">aide_predict.bespoke_models.predictors.saprot</span> <span class="kn">import</span> <span class="n">get_structure_tokens</span>
<span class="c1">#check if &#39;foldseek&#39; is available to path for a terminal</span>
<span class="k">if</span> <span class="n">shutil</span><span class="o">.</span><span class="n">which</span><span class="p">(</span><span class="s1">&#39;foldseek&#39;</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">AVAILABLE</span> <span class="o">=</span> <span class="n">MessageBool</span><span class="p">(</span><span class="kc">False</span><span class="p">,</span> <span class="s1">&#39;Foldseek is not available, please install and set environment variables.&#39;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">AVAILABLE</span> <span class="o">=</span> <span class="n">MessageBool</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="s1">&#39;Foldseek is available&#39;</span><span class="p">)</span>

<span class="k">class</span> <span class="nc">FoldseekKmerEmbedding</span><span class="p">(</span><span class="n">RequiresStructureMixin</span><span class="p">,</span> <span class="n">CanHandleAlignedSequencesMixin</span><span class="p">,</span> <span class="n">ProteinModelWrapper</span><span class="p">):</span>
    <span class="n">_available</span><span class="o">=</span><span class="n">AVAILABLE</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">metadata_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
                 <span class="n">k</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> 
                 <span class="n">wt</span><span class="p">:</span> <span class="n">ProteinSequence</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">metadata_folder</span><span class="o">=</span><span class="n">metadata_folder</span><span class="p">,</span> <span class="n">wt</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">=</span> <span class="n">k</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_kmer_to_index</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">def</span> <span class="nf">_fit</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">:</span> <span class="n">ProteinSequences</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;KmerEmbedding&#39;</span><span class="p">:</span>
        <span class="n">unique_kmers</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">X</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">seq</span><span class="o">.</span><span class="n">structure</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;KmerEmbedding requires a structure to be present in each sequence.&quot;</span><span class="p">)</span>

            <span class="n">struct_str</span> <span class="o">=</span> <span class="n">get_structure_tokens</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">structure</span><span class="p">)</span>
            <span class="n">unique_kmers</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">struct_str</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="bp">self</span><span class="o">.</span><span class="n">k</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">struct_str</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">_kmer_to_index</span> <span class="o">=</span> <span class="p">{</span><span class="n">kmer</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">kmer</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">unique_kmers</span><span class="p">))}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_features_</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_kmer_to_index</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fitted_</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="nf">_transform</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">:</span> <span class="n">ProteinSequences</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Transform the protein sequences into K-mer embeddings.</span>

<span class="sd">        Args:</span>
<span class="sd">            X (ProteinSequences): The input protein sequences.</span>

<span class="sd">        Returns:</span>
<span class="sd">            np.ndarray: The K-mer embeddings for the sequences.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">embeddings</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">X</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_features_</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">X</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">seq</span><span class="o">.</span><span class="n">structure</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;KmerEmbedding requires a structure to be present in each sequence.&quot;</span><span class="p">)</span>
            
            <span class="n">struct_str</span> <span class="o">=</span> <span class="n">get_structure_tokens</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">structure</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">struct_str</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                <span class="n">kmer</span> <span class="o">=</span> <span class="n">struct_str</span><span class="p">[</span><span class="n">j</span><span class="p">:</span><span class="n">j</span><span class="o">+</span><span class="bp">self</span><span class="o">.</span><span class="n">k</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">kmer</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_kmer_to_index</span><span class="p">:</span>
                    <span class="n">embeddings</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_kmer_to_index</span><span class="p">[</span><span class="n">kmer</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">return</span> <span class="n">embeddings</span>
</pre></div>
</div>
<p>Here we define an arbitrary predictor that calls a third party script and environment and communicates with aide via IO. We can check for model availability by checking for enviornment variables associated with the third party environment and location if necessary.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Optional</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">from</span> <span class="nn">aide_predict.bespoke_models.base</span> <span class="kn">import</span> <span class="n">ProteinModelWrapper</span><span class="p">,</span> <span class="n">RequiresStructureMixin</span><span class="p">,</span> <span class="n">RequiresFixedLengthSequencesMixin</span>
<span class="kn">from</span> <span class="nn">aide_predict.utils.data_structures</span> <span class="kn">import</span> <span class="n">ProteinSequences</span><span class="p">,</span> <span class="n">ProteinSequence</span>
<span class="kn">from</span> <span class="nn">aide_predict.utils.common</span> <span class="kn">import</span> <span class="n">MessageBool</span>
<span class="kn">from</span> <span class="nn">aide_predict.bespoke_models.predictors.saprot</span> <span class="kn">import</span> <span class="n">get_structure_tokens</span>

<span class="k">try</span><span class="p">:</span>
    <span class="n">ENV_NAME</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;BESPOKE_ENV_NAME&#39;</span><span class="p">]</span>
    <span class="n">MODEL_BASE_DIR</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;BESPOKE_MODEL_BASE_DIR&#39;</span><span class="p">]</span>
    <span class="n">AVAILABLE</span> <span class="o">=</span> <span class="n">MessageBool</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="s1">&#39;Model is available&#39;</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
    <span class="n">AVAILABLE</span> <span class="o">=</span> <span class="n">MessageBool</span><span class="p">(</span><span class="kc">False</span><span class="p">,</span> <span class="s1">&#39;Model is not available, please install and set environment variables.&#39;</span><span class="p">)</span>

<span class="k">class</span> <span class="nc">MyBespokeModel</span><span class="p">(</span><span class="n">ProteinModelWrapper</span><span class="p">,</span> <span class="n">RequiresStructureMixin</span><span class="p">,</span> <span class="n">RequiresFixedLengthSequencesMixin</span><span class="p">):</span>
    <span class="n">_available</span> <span class="o">=</span> <span class="n">AVAILABLE</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="c1"># set params....</span>

    <span class="k">def</span> <span class="nf">_fit</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">:</span> <span class="n">ProteinSequences</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="c1"># prepare data</span>
        <span class="c1"># call subprocess</span>
        <span class="n">input_fasta</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">delete</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">X</span><span class="o">.</span><span class="n">to_fasta</span><span class="p">(</span><span class="n">input_fasta</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="c1"># temdir for structures</span>
        <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">()</span> <span class="k">as</span> <span class="n">tempdir</span><span class="p">:</span>
            <span class="c1"># pass structure files to the script</span>
            <span class="n">structure_files</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tempdir</span><span class="p">,</span> <span class="s1">&#39;structure_files.txt&#39;</span><span class="p">),</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">X</span><span class="p">:</span>
                <span class="n">struct_path</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">pdb_file</span>
                <span class="n">structure_files</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">seq</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">struct_path</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">structure_files</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

            <span class="c1"># call subprocess in the environment and base dir</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;python&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/train_script.py&#39;</span><span class="p">,</span> <span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="n">structure_files</span><span class="p">,</span> <span class="n">input_fasta</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata_folder</span><span class="p">,</span> <span class="s1">&#39;model.pkl&#39;</span><span class="p">),</span> <span class="n">env</span><span class="o">=</span><span class="n">ENV_NAME</span><span class="p">,</span> <span class="n">cwd</span><span class="o">=</span><span class="n">MODEL_BASE_DIR</span><span class="p">])</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="nf">_transform</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">:</span> <span class="n">ProteinSequences</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
        <span class="c1"># something similar to fit, call a predict script etc...</span>

        <span class="k">return</span> <span class="n">outputs</span><span class="o">.</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installing AIDE" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="data_structures.html" class="btn btn-neutral float-right" title="Data Structures" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Gregg T. Beckham.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>