

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Data Structures &mdash; aide 1.1.01 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=e50f1ef7"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Model Compatibility" href="model_compatibility.html" />
    <link rel="prev" title="API examples" href="api_examples.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            aide
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installing AIDE</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_examples.html">API examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Data Structures</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#proteinsequence">ProteinSequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="#proteinsequences">ProteinSequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="#proteinstructure">ProteinStructure</a></li>
<li class="toctree-l2"><a class="reference internal" href="#structuremapper">StructureMapper</a></li>
<li class="toctree-l2"><a class="reference internal" href="#proteinsequencesonfile">ProteinSequencesOnFile</a></li>
<li class="toctree-l2"><a class="reference internal" href="#proteintrajectory">ProteinTrajectory</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="model_compatibility.html">Model Compatibility</a></li>
<li class="toctree-l1"><a class="reference internal" href="protein_model.html">ProteinModelWrapper</a></li>
<li class="toctree-l1"><a class="reference internal" href="zero_shot.html">Zero-Shot Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="supervised.html">Supervised Learning</a></li>
<li class="toctree-l1"><a class="reference internal" href="saturation_mutagenesis.html">Saturation Mutagenesis</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipelines.html">Building ML Pipelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="caching.html">Caching Model Outputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="position_specific.html">Position-Specific Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing_models.html">Contributing Models to AIDE</a></li>
<li class="toctree-l1"><a class="reference internal" href="structure_pred.html">Structure Prediction with SoloSeq</a></li>
<li class="toctree-l1"><a class="reference internal" href="msa_search.html">Generating MSAs with MMseqs2</a></li>
<li class="toctree-l1"><a class="reference internal" href="badass.html">Protein Optimization with BADASS</a></li>
<li class="toctree-l1"><a class="reference internal" href="roadmap.html">Roadmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="resource_test.html">Resource testing</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules.html">aide_predict</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">aide</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Data Structures</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/user_guide/data_structures.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section class="tex2jax_ignore mathjax_ignore" id="data-structures">
<h1>Data Structures<a class="headerlink" href="#data-structures" title="Link to this heading"></a></h1>
<p>AIDE provides several key data structures for working with protein sequences and structures. Models will receive these data structures as input in analog to how sklearn receives numpy arrays.</p>
<section id="proteinsequence">
<h2>ProteinSequence<a class="headerlink" href="#proteinsequence" title="Link to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">ProteinSequence</span></code> is the basic unit for representing a protein sequence. It behaves like a string but provides additional functionality specific to protein sequences.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">aide_predict</span> <span class="kn">import</span> <span class="n">ProteinSequence</span>

<span class="c1"># Create a basic sequence</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">ProteinSequence</span><span class="p">(</span><span class="s2">&quot;MKLLVLGLPGAGKGT&quot;</span><span class="p">)</span>

<span class="c1"># or get from a pdb (see ProteinStructure below)</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">ProteinSequence</span><span class="o">.</span><span class="n">from_pdb</span><span class="p">(</span><span class="s2">&quot;path/to/structure.pdb&quot;</span><span class="p">)</span>

<span class="c1"># Add identifier and optional structure</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">ProteinSequence</span><span class="p">(</span>
    <span class="s2">&quot;MKLLVLGLPGAGKGT&quot;</span><span class="p">,</span>
    <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;P1234&quot;</span><span class="p">,</span>
    <span class="n">structure</span><span class="o">=</span><span class="s2">&quot;path/to/structure.pdb&quot;</span>
<span class="p">)</span>

<span class="c1"># Create a sequence with an associated MSA</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">ProteinSequence</span><span class="p">(</span>
    <span class="s2">&quot;MKLLVLGLPGAGKGT&quot;</span><span class="p">,</span>
    <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;P1234&quot;</span><span class="p">,</span>
    <span class="n">msa</span><span class="o">=</span><span class="n">msa_sequences</span>  <span class="c1"># ProteinSequences object containing the MSA</span>
<span class="p">)</span>

<span class="c1"># From a fasta file, assumed to be aligned. Uses first sequence in file, set&#39;s the MSA attribute</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">ProteinSequence</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="s2">&quot;path/to/aligned.fasta&quot;</span><span class="p">)</span>
<span class="n">seq</span><span class="o">.</span><span class="n">has_msa</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kc">True</span>
</pre></div>
</div>
<p>Key attributes and methods:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">id</span></code>: Optional identifier for the sequence</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">structure</span></code>: Optional associated structure (as ProteinStructure object)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">msa</span></code>: Optional associated multiple sequence alignment (as ProteinSequences object)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">has_gaps</span></code>: Boolean indicating if sequence contains gaps (‘-’ or ‘.’)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">base_length</span></code>: Length excluding gaps</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">has_non_canonical</span></code>: Boolean indicating presence of non-standard amino acids</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">as_array</span></code>: Sequence as numpy array</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">has_msa</span></code>: Boolean indicating if the sequence has an associated MSA</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">msa_same_width</span></code>: Boolean indicating if the MSA has the same width as the sequence</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">is_in_msa</span></code>: Boolean indicating if the sequence is present in its MSA</p></li>
</ul>
<p>Common operations:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Sequence manipulation</span>
<span class="n">ungapped</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">with_no_gaps</span><span class="p">()</span>  <span class="c1"># Remove gaps</span>
<span class="n">upper_seq</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>  <span class="c1"># Convert to uppercase</span>

<span class="c1"># Mutation operations</span>
<span class="n">mutated</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">mutate</span><span class="p">(</span><span class="s2">&quot;A123G&quot;</span><span class="p">)</span>  <span class="c1"># Single mutation</span>
<span class="n">mutated</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">mutate</span><span class="p">([</span><span class="s2">&quot;A123G&quot;</span><span class="p">,</span> <span class="s2">&quot;L45R&quot;</span><span class="p">])</span>  <span class="c1"># Multiple mutations</span>

<span class="c1"># Compare sequences</span>
<span class="n">positions</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">mutated_positions</span><span class="p">(</span><span class="n">other_seq</span><span class="p">)</span>  <span class="c1"># Get positions that differ</span>
<span class="n">mutations</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">get_mutations</span><span class="p">(</span><span class="n">other_seq</span><span class="p">)</span>  <span class="c1"># Get mutation strings (e.g., &quot;A123G&quot;)</span>

<span class="c1"># Create all possible mutations</span>
<span class="n">library</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">saturation_mutagenesis</span><span class="p">()</span>  <span class="c1"># Generate all single mutants</span>
<span class="n">library</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">saturation_mutagenesis</span><span class="p">(</span><span class="n">positions</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span>  <span class="c1"># Specific positions</span>

<span class="c1"># align to another sequence</span>
<span class="n">seq2</span> <span class="o">=</span> <span class="n">seq2</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

<span class="c1"># align to an existing alignment</span>
<span class="n">seq2</span> <span class="o">=</span> <span class="n">seq2</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span>  <span class="c1"># Align to existing MSA</span>

</pre></div>
</div>
</section>
<section id="proteinsequences">
<h2>ProteinSequences<a class="headerlink" href="#proteinsequences" title="Link to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">ProteinSequences</span></code> manages collections of protein sequences, similar to a list but with additional functionality for protein-specific operations.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">aide_predict</span> <span class="kn">import</span> <span class="n">ProteinSequences</span>

<span class="c1"># Create from various sources</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="p">([</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">seq3</span><span class="p">])</span>  <span class="c1"># From ProteinSequence objects</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="s2">&quot;sequences.fasta&quot;</span><span class="p">)</span>  <span class="c1"># From FASTA file</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_a3m</span><span class="p">(</span><span class="s2">&quot;alignment.a3m&quot;</span><span class="p">)</span>  <span class="c1"># From A3M file</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_list</span><span class="p">([</span><span class="s2">&quot;MKLL...&quot;</span><span class="p">,</span> <span class="s2">&quot;MKLT...&quot;</span><span class="p">])</span>  <span class="c1"># From strings</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;MKLL...&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;MKLT...&quot;</span><span class="p">})</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">my_dataframe</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">())</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_df</span><span class="p">(</span><span class="n">my_dataframe</span><span class="p">)</span> <span class="c1"># assumes first column is sequences</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_df</span><span class="p">(</span><span class="n">my_dataframe</span><span class="p">,</span> <span class="n">sequence_col</span><span class="o">=</span><span class="s2">&quot;seq_col&quot;</span><span class="p">,</span> <span class="n">id_col</span><span class="o">=</span><span class="s2">&quot;id_col&quot;</span><span class="p">)</span> <span class="c1"># specify columns</span>
<span class="n">sequences</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_df</span><span class="p">(</span><span class="n">my_dataframe</span><span class="p">,</span> <span class="n">label_cols</span><span class="o">=</span><span class="s1">&#39;label&#39;</span><span class="p">)</span> <span class="c1"># get a array of labels</span>
</pre></div>
</div>
<p>Key attributes:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">aligned</span></code>: Boolean indicating if all sequences have same length (including gaps)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fixed_length</span></code>: Boolean indicating if all sequences have same base length (excluding gaps)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">width</span></code>: Length of sequences if aligned, None otherwise</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">has_gaps</span></code>: Boolean indicating if any sequence contains gaps</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mutated_positions</span></code>: List of positions with variation (for aligned sequences)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">weights</span></code>: Optional weights for each sequence (used in some models)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ids</span></code>: List of sequence IDs</p></li>
</ul>
<p>Common operations:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Sequence alignment</span>
<span class="n">aligned</span> <span class="o">=</span> <span class="n">sequences</span><span class="o">.</span><span class="n">align_all</span><span class="p">()</span>  <span class="c1"># Align all sequences</span>
<span class="n">aligned</span> <span class="o">=</span> <span class="n">sequences</span><span class="o">.</span><span class="n">align_to</span><span class="p">(</span><span class="n">reference_msa</span><span class="p">)</span>  <span class="c1"># Align to existing MSA</span>

<span class="c1"># Access and manipulation</span>
<span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># Access by index</span>
<span class="n">sequences</span><span class="p">[</span><span class="s2">&quot;seq1&quot;</span><span class="p">]</span>  <span class="c1"># Access by ID</span>
<span class="n">sequences</span><span class="o">.</span><span class="n">with_no_gaps</span><span class="p">()</span>  <span class="c1"># Remove gaps from all sequences</span>
<span class="n">sequences</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>  <span class="c1"># Convert all to uppercase</span>

<span class="c1"># MSA operations</span>
<span class="n">msa</span> <span class="o">=</span> <span class="n">sequences</span><span class="o">.</span><span class="n">msa_process</span><span class="p">(</span>
    <span class="n">focus_seq_id</span><span class="o">=</span><span class="s2">&quot;wild_type&quot;</span><span class="p">,</span>
    <span class="n">theta</span><span class="o">=</span><span class="mf">0.8</span>  <span class="c1"># Sequence reweighting parameter</span>
<span class="p">)</span>

<span class="c1"># Sampling operations</span>
<span class="n">sampled</span> <span class="o">=</span> <span class="n">sequences</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">keep_first</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># Sample sequences, keeping the first one</span>

<span class="c1"># Save/export</span>
<span class="n">sequences</span><span class="o">.</span><span class="n">to_fasta</span><span class="p">(</span><span class="s2">&quot;output.fasta&quot;</span><span class="p">)</span>
<span class="n">sequences_dict</span> <span class="o">=</span> <span class="n">sequences</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>

<span class="c1"># Batching for large datasets</span>
<span class="k">for</span> <span class="n">batch</span> <span class="ow">in</span> <span class="n">sequences</span><span class="o">.</span><span class="n">iter_batches</span><span class="p">(</span><span class="n">batch_size</span><span class="o">=</span><span class="mi">32</span><span class="p">):</span>
    <span class="c1"># Process batch</span>
    <span class="k">pass</span>

<span class="c1"># Get mapping between aligned and original positions</span>
<span class="k">if</span> <span class="n">sequences</span><span class="o">.</span><span class="n">aligned</span><span class="p">:</span>
    <span class="n">alignment_mapping</span> <span class="o">=</span> <span class="n">sequences</span><span class="o">.</span><span class="n">get_alignment_mapping</span><span class="p">()</span>
    <span class="c1"># Returns dict mapping sequence IDs to lists of positions</span>

<span class="c1"># Convert to on-file representation for large MSAs</span>
<span class="n">on_file</span> <span class="o">=</span> <span class="n">sequences</span><span class="o">.</span><span class="n">to_on_file</span><span class="p">(</span><span class="s2">&quot;output.fasta&quot;</span><span class="p">)</span>  <span class="c1"># Save to file and return a file-based object</span>
</pre></div>
</div>
</section>
<section id="proteinstructure">
<h2>ProteinStructure<a class="headerlink" href="#proteinstructure" title="Link to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">ProteinStructure</span></code> represents the 3D structure of a protein, integrating with common structure file formats and analysis tools.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">aide_predict</span> <span class="kn">import</span> <span class="n">ProteinStructure</span>

<span class="c1"># Create from PDB file</span>
<span class="n">structure</span> <span class="o">=</span> <span class="n">ProteinStructure</span><span class="p">(</span>
    <span class="n">pdb_file</span><span class="o">=</span><span class="s2">&quot;protein.pdb&quot;</span><span class="p">,</span>
    <span class="n">chain</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span>  <span class="c1"># Optional chain identifier</span>
    <span class="n">plddt_file</span><span class="o">=</span><span class="s2">&quot;confidence.json&quot;</span>  <span class="c1"># Optional AlphaFold2 confidence scores</span>
<span class="p">)</span>

<span class="c1"># Create from AlphaFold2 output folder</span>
<span class="n">structure</span> <span class="o">=</span> <span class="n">ProteinStructure</span><span class="o">.</span><span class="n">from_af2_folder</span><span class="p">(</span>
    <span class="n">folder_path</span><span class="o">=</span><span class="s2">&quot;af2_results&quot;</span><span class="p">,</span>
    <span class="n">chain</span><span class="o">=</span><span class="s2">&quot;A&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
<p>Key methods:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Access structure information</span>
<span class="n">sequence</span> <span class="o">=</span> <span class="n">structure</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">()</span>  <span class="c1"># Get amino acid sequence</span>
<span class="n">plddt</span> <span class="o">=</span> <span class="n">structure</span><span class="o">.</span><span class="n">get_plddt</span><span class="p">()</span>  <span class="c1"># Get pLDDT confidence scores</span>
<span class="n">dssp</span> <span class="o">=</span> <span class="n">structure</span><span class="o">.</span><span class="n">get_dssp</span><span class="p">()</span>  <span class="c1"># Get secondary structure assignments</span>

<span class="c1"># Validation</span>
<span class="n">structure</span><span class="o">.</span><span class="n">validate_sequence</span><span class="p">(</span><span class="s2">&quot;MKLLVLGLPGAGKGT&quot;</span><span class="p">)</span>  <span class="c1"># Check if sequence matches structure</span>

<span class="c1"># Access underlying structure objects</span>
<span class="n">structure_obj</span> <span class="o">=</span> <span class="n">structure</span><span class="o">.</span><span class="n">get_structure</span><span class="p">()</span>  <span class="c1"># Get BioPython Structure object</span>
<span class="n">chain_obj</span> <span class="o">=</span> <span class="n">structure</span><span class="o">.</span><span class="n">get_chain</span><span class="p">()</span>  <span class="c1"># Get specific chain</span>
<span class="n">positions</span> <span class="o">=</span> <span class="n">structure</span><span class="o">.</span><span class="n">get_residue_positions</span><span class="p">()</span>  <span class="c1"># Get residue numbers</span>
</pre></div>
</div>
</section>
<section id="structuremapper">
<h2>StructureMapper<a class="headerlink" href="#structuremapper" title="Link to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">StructureMapper</span></code> helps manage multiple structures and map them to sequences, particularly useful when working with structure-aware models.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">aide_predict</span> <span class="kn">import</span> <span class="n">StructureMapper</span>

<span class="c1"># Initialize with folder containing structures</span>
<span class="n">mapper</span> <span class="o">=</span> <span class="n">StructureMapper</span><span class="p">(</span><span class="s2">&quot;path/to/structures&quot;</span><span class="p">)</span>

<span class="c1"># Structure can be PDB files or AlphaFold2 prediction folders</span>
<span class="c1"># Example folder structure:</span>
<span class="c1"># structures/</span>
<span class="c1">#   ├── protein1.pdb</span>
<span class="c1">#   ├── protein2.pdb</span>
<span class="c1">#   └── protein3/  # AlphaFold2 output folder</span>
<span class="c1">#       ├── ranked_0.pdb</span>
<span class="c1">#       └── ranking_confidence.json</span>

<span class="c1"># Assign structures to ProteinSequences already loaded </span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">mapper</span><span class="o">.</span><span class="n">assign_structures</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>

<span class="c1"># Get available structures</span>
<span class="n">available_ids</span> <span class="o">=</span> <span class="n">mapper</span><span class="o">.</span><span class="n">get_available_structures</span><span class="p">()</span>

<span class="c1"># Get ProteinSequences with structures</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">mapper</span><span class="o">.</span><span class="n">get_protein_sequences</span><span class="p">()</span>
</pre></div>
</div>
<p>The StructureMapper is particularly useful when working with structure-aware models like SaProt, which can use structure information to improve predictions:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example workflow with structure-aware model</span>
<span class="n">mapper</span> <span class="o">=</span> <span class="n">StructureMapper</span><span class="p">(</span><span class="s2">&quot;structures/&quot;</span><span class="p">)</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">ProteinSequences</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="s2">&quot;sequences.fasta&quot;</span><span class="p">)</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">mapper</span><span class="o">.</span><span class="n">assign_structures</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>

<span class="n">model</span> <span class="o">=</span> <span class="n">SaProtLikelihoodWrapper</span><span class="p">(</span><span class="n">wt</span><span class="o">=</span><span class="n">sequences</span><span class="p">[</span><span class="s2">&quot;wild_type&quot;</span><span class="p">])</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>  <span class="c1"># Will use structures where available, falling back to the WT structure</span>
</pre></div>
</div>
</section>
<section id="proteinsequencesonfile">
<h2>ProteinSequencesOnFile<a class="headerlink" href="#proteinsequencesonfile" title="Link to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">ProteinSequencesOnFile</span></code> provides a memory-efficient way to work with large alignments by only loading sequences when needed.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">aide_predict</span> <span class="kn">import</span> <span class="n">ProteinSequencesOnFile</span>

<span class="c1"># Create from FASTA file</span>
<span class="n">on_file_sequences</span> <span class="o">=</span> <span class="n">ProteinSequencesOnFile</span><span class="p">(</span><span class="s2">&quot;large_alignment.fasta&quot;</span><span class="p">)</span>

<span class="c1"># Access properties efficiently without loading everything</span>
<span class="nb">print</span><span class="p">(</span><span class="n">on_file_sequences</span><span class="o">.</span><span class="n">aligned</span><span class="p">)</span>  <span class="c1"># Check if aligned</span>
<span class="nb">print</span><span class="p">(</span><span class="n">on_file_sequences</span><span class="o">.</span><span class="n">width</span><span class="p">)</span>    <span class="c1"># Get width if aligned</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">on_file_sequences</span><span class="p">))</span>     <span class="c1"># Get count of sequences</span>

<span class="c1"># Load a specific sequence</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">on_file_sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># Get by index</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">on_file_sequences</span><span class="p">[</span><span class="s2">&quot;seq1&quot;</span><span class="p">]</span>  <span class="c1"># Get by ID</span>

<span class="c1"># Iterate through sequences without loading all at once</span>
<span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">on_file_sequences</span><span class="p">:</span>
    <span class="n">process_sequence</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

<span class="c1"># Load into memory if needed</span>
<span class="n">in_memory</span> <span class="o">=</span> <span class="n">on_file_sequences</span><span class="o">.</span><span class="n">to_memory</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="proteintrajectory">
<h2>ProteinTrajectory<a class="headerlink" href="#proteintrajectory" title="Link to this heading"></a></h2>
<p>NOT YET IMPLEMENTED</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="api_examples.html" class="btn btn-neutral float-left" title="API examples" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="model_compatibility.html" class="btn btn-neutral float-right" title="Model Compatibility" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Gregg T. Beckham.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>